File: desktop/enc_roller/engine/engine.py
"""engine.py — every roll mechanic.  Pure; stdlib only; no tkinter.

Design rules that keep the mechanics honest:
  * A single find-site, ``_match_or_gap``, so a ``None`` on a legal roll is
    structurally impossible to forget.
  * One shared ``depth`` counter threaded through item-rerolls, the
    Enchanted-Enhancement selector, and R3/S3 sub-cascades — the reroll cap is
    a single global budget (``REROLL_CAP`` executed combines).
  * The result is always a ``RollStep`` tree; ``_headline`` derives the
    read-aloud name from that tree, so display and mechanics never diverge.
"""

from __future__ import annotations

import re

from ..data.dataset import Dataset
from ..data.models import Item
from .config import MAX_POWERS_SOFT, POWER_RANDOM_POOL, REROLL_CAP
from .dice import Roller
from .results import RollResult, RollStep

EE_NAME = "Enchanted Enhancements"


class RollEngine:
    def __init__(self, ds: Dataset, roller: Roller, cap: int = REROLL_CAP,
                 power_pool=POWER_RANDOM_POOL):
        self.ds = ds
        self.roller = roller
        self.cap = cap
        self.pool = power_pool               # None -> all power tables

    # ===================================================================== #
    # The single find-site.  Every roll->row lookup goes through here.
    # ===================================================================== #
    def _match_or_gap(self, rows, roll, table, die, *, nearest=False) -> RollStep:
        row = self.ds.find(rows, roll)
        if row is None and nearest:
            row, note = self.ds.find_nearest(rows, roll)
            if row is not None:
                return RollStep(table, die, roll, row.name,
                                getattr(row, "page", None),
                                getattr(row, "page_status", "n/a"),
                                kind="roll",
                                note=f"data gap at {roll}; used nearest ({note})")
        if row is None:
            return RollStep(table, die, roll, "(no entry for roll — data gap)",
                            None, "n/a", kind="gap",
                            note="data gap; re-roll this slot")
        return RollStep(table, die, roll, row.name,
                        getattr(row, "page", None),
                        getattr(row, "page_status", "n/a"))

    # ===================================================================== #
    # Feature 1 — roll any single table
    # ===================================================================== #
    def roll_item_table(self, table: str, depth: int = 0) -> RollResult:
        rows = self.ds.items_by_table[table]
        roll = self.roller.roll(1000)
        step = self._match_or_gap(rows, roll, table, 1000)
        if step.kind == "roll":
            matched = self.ds.find(rows, roll)
            self._attach_cascade(step, matched, table, depth)
        return RollResult("single", self._headline(step), step, self.roller.seed)

    def roll_power_table(self, num: str) -> RollResult:
        return self.roll_artifact_power(num)

    def roll_mech(self, key: str) -> RollResult:
        """Roll one raw mechanics sub-table (R1/R2/S1/S2) on its own."""
        die = 1000 if key in ("R1", "S1") else 20
        roll = self.roller.roll(die)
        step = self._match_or_gap(self.ds.mech[key], roll, key, die,
                                  nearest=(key in ("R1", "S1")))
        return RollResult("single", step.label, step, self.roller.seed)

    def roll_armor(self, depth: int = 0) -> RollResult:
        root = self._assemble_armor(depth)
        return RollResult("single", self._headline(root), root, self.roller.seed)

    def roll_weapon(self, depth: int = 0) -> RollResult:
        root = self._assemble_weapon(depth)
        return RollResult("single", self._headline(root), root, self.roller.seed)

    def roll_named(self, key: str) -> RollResult:
        """UI dispatch by combo id."""
        if key in self.ds.items_by_table:
            return self.roll_item_table(key)
        if key == "R":
            return self.roll_armor()
        if key == "S":
            return self.roll_weapon()
        if key in self.ds.mech:
            return self.roll_mech(key)
        if key in self.ds.power_by_num:
            return self.roll_power_table(key)
        raise KeyError(f"unknown roll target {key!r}")

    def _attach_cascade(self, step: RollStep, item, table: str, depth: int) -> None:
        if item is None:
            return
        if item.name == EE_NAME:
            step.children = [self._resolve_enchanted_enhancements(depth + 1)]
        elif item.reroll:
            step.children = [self._resolve_item(item, table, depth + 1)]

    # ===================================================================== #
    # Feature 2 — fully random item (master -> target, with R/S assembly)
    # ===================================================================== #
    def roll_random_item(self) -> RollResult:
        d = self.roller.roll(100)
        mrow = self.ds.find(self.ds.master_rows, d)
        if mrow is None:                              # defensive: master tiles fully
            hi = max((r.roll_high for r in self.ds.master_rows), default=0)
            root = RollStep("master", 100, d,
                            f"No category — source covers d100 1-{hi}",
                            None, "n/a", kind="gap", note="master gap")
            return RollResult("random_item", "No result — master gap", root,
                              self.roller.seed)
        if mrow.target == "R":
            inner = self._assemble_armor(0)
        elif mrow.target == "S":
            inner = self._assemble_weapon(0)
        else:
            inner = self.roll_item_table(mrow.target, 0).root
        root = RollStep("master", 100, d, mrow.name, None, "n/a", children=[inner])
        return RollResult("random_item", self._headline(root), root, self.roller.seed)

    # ===================================================================== #
    # Feature 3 & 4 — artifact powers + N-power artifacts
    # ===================================================================== #
    def roll_artifact_power(self, num: str | None = None) -> RollResult:
        pt = self._pick_power_table(num)
        r = self.roller.roll(pt.die)
        step = self._match_or_gap(pt.entries, r, f"power:{pt.num}", pt.die)
        if step.note is None:
            step.note = f"{pt.category} ({pt.num})"
        return RollResult("artifact_power", step.label, step, self.roller.seed)

    def generate_artifact(self, n: int, tables=None) -> RollResult:
        if not isinstance(n, int) or n < 1:
            raise ValueError("N must be an integer >= 1")
        if tables is not None and len(tables) != n:
            raise ValueError("tables length must equal N")
        picks = tables if tables is not None else [None] * n
        children = []
        for t in picks:
            child = self.roll_artifact_power(t).root
            if child.kind == "gap":                  # a data hole -> re-roll the slot once
                child = self.roll_artifact_power(t).root
            children.append(child)
        root = RollStep("artifact", 0, 0, f"Artifact — {n} powers", None, "n/a",
                        kind="assembly", children=children)
        if n > MAX_POWERS_SOFT:
            root.note = f"large artifact (N={n} > soft cap {MAX_POWERS_SOFT})"
        return RollResult("artifact", self._headline(root), root, self.roller.seed)

    def _pick_power_table(self, num):
        if num is not None:
            return self.ds.power_by_num[num]
        pool = self.pool or sorted(self.ds.power_by_num)
        idx = self.roller.roll(len(pool)) - 1
        return self.ds.power_by_num[pool[idx]]

    # ===================================================================== #
    # Bonus — treasure hoard
    # ===================================================================== #
    def roll_hoard(self, k: int) -> RollResult:
        if not isinstance(k, int) or k < 1:
            raise ValueError("hoard size must be an integer >= 1")
        children = [self.roll_random_item().root for _ in range(k)]
        root = RollStep("hoard", 0, 0, f"Treasure Hoard — {k} items", None, "n/a",
                        kind="assembly", children=children)
        return RollResult("hoard", self._headline(root), root, self.roller.seed)

    # ===================================================================== #
    # Shared primitives — cascade, EE selector, R/S assembly
    # ===================================================================== #
    def _resolve_item(self, item: Item, table: str, depth: int) -> RollStep:
        """Execute one reroll/combine.  Invoked because ``item.reroll``; depth>=1."""
        if item.name == EE_NAME:                     # defensive: EE routes elsewhere already
            return self._resolve_enchanted_enhancements(depth)
        if depth > self.cap:
            return RollStep(table, 0, 0, "(reroll cap reached)", None, "n/a",
                            kind="cap", note=f"cap ({self.cap}) reached")
        rows = self.ds.items_by_table[table]
        roll = self.roller.roll(1000)
        step = self._match_or_gap(rows, roll, table, 1000)
        if step.kind != "roll":                      # a gap during reroll: leave it, no combine
            return step
        step.kind = "reroll"
        step.note = "reroll → combined"
        matched = self.ds.find(rows, roll)
        if matched is not None and matched.name == EE_NAME:
            step.note = "reroll → Enchanted Enhancements"
            step.children = [self._resolve_enchanted_enhancements(depth + 1)]
        elif matched is not None and matched.reroll:
            step.children = [self._resolve_item(matched, table, depth + 1)]
        return step

    def _resolve_enchanted_enhancements(self, depth: int) -> RollStep:
        if depth > self.cap:
            return RollStep("enh", 0, 0, "(reroll cap reached)", None, "n/a",
                            kind="cap", note=f"cap ({self.cap}) reached")
        d = self.roller.roll(100)
        row = self.ds.find(self.ds.enhancement, d)
        if row is None:
            return RollStep("enh", 100, d, "(no enhancement for roll — data gap)",
                            None, "n/a", kind="gap", note="data gap")
        step = RollStep("enh", 100, d, f"Enchanted: {row.type}", None, "n/a",
                        kind="enhancement")
        if row.reroll:
            step.children = [self._resolve_enchanted_enhancements(depth + 1)]
        return step

    def _assemble_armor(self, depth: int = 0) -> RollStep:
        return self._assemble_rs("armor", depth)

    def _assemble_weapon(self, depth: int = 0) -> RollStep:
        return self._assemble_rs("weapon", depth)

    def _assemble_rs(self, slot: str, depth: int) -> RollStep:
        if slot == "armor":
            t1, t2, item_table, label = "R1", "R2", "R3", "Armor"
        else:
            t1, t2, item_table, label = "S1", "S2", "S3", "Weapon"

        t = self.roller.roll(1000)
        type_step = self._match_or_gap(self.ds.mech[t1], t, t1, 1000, nearest=True)
        matched_type = self.ds.find(self.ds.mech[t1], t)

        if matched_type is not None and matched_type.is_r3_catchall:
            # Special -> roll the specific item table (cascades); no generic bonus.
            type_step.note = f"Special → roll on {item_table}"
            item_step = self.roll_item_table(item_table, depth).root   # shares the outer budget
            return RollStep(slot, 0, 0, f"{label} (Special)", None, "n/a",
                            kind="assembly", children=[type_step, item_step],
                            note=f"{t1} Special → specific item from {item_table}")

        # Generic magic armor/weapon: type + bonus (no R3/S3 roll).
        b = self.roller.roll(20)
        bonus_step = self._match_or_gap(self.ds.mech[t2], b, t2, 20)
        if slot == "weapon" and matched_type is not None and matched_type.name != "Sword":
            matched_bonus = self.ds.find(self.ds.mech[t2], b)          # S2 Wpn Adj for non-swords
            if matched_bonus is not None and getattr(matched_bonus, "wpn_adj", None):
                bonus_step.label = matched_bonus.wpn_adj
                bonus_step.note = "Wpn Adj (non-sword)"
        return RollStep(slot, 0, 0, label, None, "n/a", kind="assembly",
                        children=[type_step, bonus_step])

    def _bonus_plus(self, mech_key: str, name: str) -> str:
        """Sign-preserving bonus token; per-table because R2/S2 name formats differ."""
        if mech_key == "S2":                         # S2 name IS the bonus: "-1".."+5"
            return name.strip()
        m = re.search(r"AC Adj\s*([+-]?\d+)", name)  # R2: extract the AC Adj token
        if not m:
            return name                              # fallback: raw name, never crash
        g = m.group(1)
        return ("+" + g) if g[0] not in "+-" else g

    # ===================================================================== #
    # Headline derivation (from the tree)
    # ===================================================================== #
    def _headline(self, root: RollStep) -> str:
        base = self._primary_name(root)
        if root.table in ("artifact", "hoard"):
            return base
        combines = root.count_kinds({"reroll", "enhancement"})
        ench = root.count_kinds({"enhancement"})
        if combines == 0:
            return base
        suffix = f"  ·  +{combines} combined" + (f" ({ench} enchanted)" if ench else "")
        return base + suffix

    def _primary_name(self, root: RollStep) -> str:
        if root.table in ("armor", "weapon"):
            return self._assembled_name(root)
        if root.table == "master" and root.children:
            return self._primary_name(root.children[0])
        return root.label

    def _assembled_name(self, root: RollStep) -> str:
        type_step, second = root.children[0], root.children[1]
        if second.table in ("R2", "S2"):        # generic: "{bonus} {type}"
            plus = self._bonus_plus(second.table, second.label)
            return f"{plus} {type_step.label}".strip()
        # Special: the R3/S3 item IS the result (bonus predetermined on its page)
        return self._primary_name(second)

File: web/js/engine.js
/* engine.js — every roll mechanic (faithful port of enc_roller/engine/engine.py).
   Single find-site (matchOrGap); one shared depth cap across item-rerolls, the
   Enchanted-Enhancement selector, and R3/S3 sub-cascades; returns a RollStep tree. */
(function () {
  "use strict";
  const EM = (globalThis.EM = globalThis.EM || {});

  const EE_NAME = "Enchanted Enhancements";
  EM.REROLL_CAP = 3;
  EM.MAX_POWERS_SOFT = 20;

  function step(table, die, rolled, label, page, page_status, extra) {
    const s = {
      table: table, die: die, rolled: rolled, label: label,
      page: (page === undefined ? null : page),
      page_status: page_status || "n/a",
      children: [], kind: "roll", note: null,
    };
    if (extra) Object.assign(s, extra);
    return s;
  }

  EM.walk = function* (node) {
    yield node;
    for (const c of node.children) yield* EM.walk(c);
  };
  EM.countKinds = function (node, kinds) {
    let n = 0;
    for (const x of EM.walk(node)) if (kinds.has(x.kind)) n++;
    return n;
  };

  EM.RollEngine = class RollEngine {
    constructor(ds, roller, opts) {
      opts = opts || {};
      this.ds = ds;
      this.roller = roller;
      this.cap = opts.cap == null ? EM.REROLL_CAP : opts.cap;
      this.pool = opts.pool || null;              // null -> all power tables
    }

    // ---- single find-site ------------------------------------------------
    matchOrGap(rows, roll, table, die, nearest) {
      let row = this.ds.find(rows, roll);
      if (row === null && nearest) {
        const [nr, note] = this.ds.findNearest(rows, roll);
        if (nr) {
          return step(table, die, roll, nr.name, nr.page, nr.page_status || "n/a",
            { kind: "roll", note: `data gap at ${roll}; used nearest (${note})` });
        }
      }
      if (row === null) {
        return step(table, die, roll, "(no entry for roll — data gap)", null, "n/a",
          { kind: "gap", note: "data gap; re-roll this slot" });
      }
      return step(table, die, roll, row.name, row.page, row.page_status || "n/a");
    }

    // ---- Feature 1: single table ----------------------------------------
    rollItemTable(table, depth) {
      depth = depth || 0;
      const rows = this.ds.itemsByTable[table];
      const roll = this.roller.roll(1000);
      const st = this.matchOrGap(rows, roll, table, 1000);
      if (st.kind === "roll") {
        const matched = this.ds.find(rows, roll);
        this.attachCascade(st, matched, table, depth);
      }
      return { kind: "single", headline: this.headline(st), root: st, seed: this.roller.seed };
    }

    rollPowerTable(num) { return this.rollArtifactPower(num); }

    rollMech(key) {
      const die = (key === "R1" || key === "S1") ? 1000 : 20;
      const roll = this.roller.roll(die);
      const st = this.matchOrGap(this.ds.mech[key], roll, key, die, key === "R1" || key === "S1");
      return { kind: "single", headline: st.label, root: st, seed: this.roller.seed };
    }

    rollArmor(depth) { const r = this.assembleArmor(depth || 0); return { kind: "single", headline: this.headline(r), root: r, seed: this.roller.seed }; }
    rollWeapon(depth) { const r = this.assembleWeapon(depth || 0); return { kind: "single", headline: this.headline(r), root: r, seed: this.roller.seed }; }

    rollNamed(key) {
      if (this.ds.itemsByTable[key]) return this.rollItemTable(key);
      if (key === "R") return this.rollArmor();
      if (key === "S") return this.rollWeapon();
      if (this.ds.mech[key]) return this.rollMech(key);
      if (this.ds.powerByNum[key]) return this.rollPowerTable(key);
      throw new Error("unknown roll target " + key);
    }

    attachCascade(st, item, table, depth) {
      if (!item) return;
      if (item.name === EE_NAME) st.children = [this.resolveEnchantedEnhancements(depth + 1)];
      else if (item.reroll) st.children = [this.resolveItem(item, table, depth + 1)];
    }

    // ---- Feature 2: fully random item -----------------------------------
    rollRandomItem() {
      const d = this.roller.roll(100);
      const mrow = this.ds.find(this.ds.masterRows, d);
      if (mrow === null) {
        const hi = this.ds.masterRows.reduce((m, r) => Math.max(m, r.roll_high), 0);
        const root = step("master", 100, d, `No category — source covers d100 1-${hi}`,
          null, "n/a", { kind: "gap", note: "master gap" });
        return { kind: "random_item", headline: "No result — master gap", root: root, seed: this.roller.seed };
      }
      let inner;
      if (mrow.target === "R") inner = this.assembleArmor(0);
      else if (mrow.target === "S") inner = this.assembleWeapon(0);
      else inner = this.rollItemTable(mrow.target, 0).root;
      const root = step("master", 100, d, mrow.name, null, "n/a", { children: [inner] });
      return { kind: "random_item", headline: this.headline(root), root: root, seed: this.roller.seed };
    }

    // ---- Features 3 & 4: artifact powers --------------------------------
    rollArtifactPower(num) {
      const pt = this.pickPowerTable(num);
      const r = this.roller.roll(pt.die);
      const st = this.matchOrGap(pt.entries, r, "power:" + pt.num, pt.die);
      if (st.note === null) st.note = `${pt.category} (${pt.num})`;
      return { kind: "artifact_power", headline: st.label, root: st, seed: this.roller.seed, category: pt.category, num: pt.num };
    }

    generateArtifact(n, tables) {
      if (!Number.isInteger(n) || n < 1) throw new Error("N must be an integer >= 1");
      if (tables != null && tables.length !== n) throw new Error("tables length must equal N");
      const picks = tables != null ? tables : new Array(n).fill(null);
      const children = [];
      for (const t of picks) {
        let child = this.rollArtifactPower(t).root;
        if (child.kind === "gap") child = this.rollArtifactPower(t).root;   // re-roll a hole once
        children.push(child);
      }
      const root = step("artifact", 0, 0, `Artifact — ${n} powers`, null, "n/a",
        { kind: "assembly", children: children });
      if (n > EM.MAX_POWERS_SOFT) root.note = `large artifact (N=${n} > soft cap ${EM.MAX_POWERS_SOFT})`;
      return { kind: "artifact", headline: this.headline(root), root: root, seed: this.roller.seed };
    }

    pickPowerTable(num) {
      if (num != null) return this.ds.powerByNum[num];
      const pool = this.pool || Object.keys(this.ds.powerByNum).sort();
      const idx = this.roller.roll(pool.length) - 1;
      return this.ds.powerByNum[pool[idx]];
    }

    // ---- Bonus: treasure hoard ------------------------------------------
    rollHoard(k) {
      if (!Number.isInteger(k) || k < 1) throw new Error("hoard size must be an integer >= 1");
      const children = [];
      for (let i = 0; i < k; i++) children.push(this.rollRandomItem().root);
      const root = step("hoard", 0, 0, `Treasure Hoard — ${k} items`, null, "n/a",
        { kind: "assembly", children: children });
      return { kind: "hoard", headline: this.headline(root), root: root, seed: this.roller.seed };
    }

    // ---- shared primitives ----------------------------------------------
    resolveItem(item, table, depth) {
      if (item.name === EE_NAME) return this.resolveEnchantedEnhancements(depth);
      if (depth > this.cap) {
        return step(table, 0, 0, "(reroll cap reached)", null, "n/a",
          { kind: "cap", note: `cap (${this.cap}) reached` });
      }
      const rows = this.ds.itemsByTable[table];
      const roll = this.roller.roll(1000);
      const st = this.matchOrGap(rows, roll, table, 1000);
      if (st.kind !== "roll") return st;
      st.kind = "reroll";
      st.note = "reroll → combined";
      const matched = this.ds.find(rows, roll);
      if (matched && matched.name === EE_NAME) {
        st.note = "reroll → Enchanted Enhancements";
        st.children = [this.resolveEnchantedEnhancements(depth + 1)];
      } else if (matched && matched.reroll) {
        st.children = [this.resolveItem(matched, table, depth + 1)];
      }
      return st;
    }

    resolveEnchantedEnhancements(depth) {
      if (depth > this.cap) {
        return step("enh", 0, 0, "(reroll cap reached)", null, "n/a",
          { kind: "cap", note: `cap (${this.cap}) reached` });
      }
      const d = this.roller.roll(100);
      const row = this.ds.find(this.ds.enhancement, d);
      if (!row) {
        return step("enh", 100, d, "(no enhancement for roll — data gap)", null, "n/a",
          { kind: "gap", note: "data gap" });
      }
      const st = step("enh", 100, d, "Enchanted: " + row.type, null, "n/a", { kind: "enhancement" });
      if (row.reroll) st.children = [this.resolveEnchantedEnhancements(depth + 1)];
      return st;
    }

    assembleArmor(depth) { return this.assembleRS("armor", depth || 0); }
    assembleWeapon(depth) { return this.assembleRS("weapon", depth || 0); }

    assembleRS(slot, depth) {
      let t1, t2, itemTable, label;
      if (slot === "armor") { t1 = "R1"; t2 = "R2"; itemTable = "R3"; label = "Armor"; }
      else { t1 = "S1"; t2 = "S2"; itemTable = "S3"; label = "Weapon"; }

      const t = this.roller.roll(1000);
      const typeStep = this.matchOrGap(this.ds.mech[t1], t, t1, 1000, true);
      const matchedType = this.ds.find(this.ds.mech[t1], t);

      if (matchedType && matchedType.is_r3_catchall) {
        // Special -> roll the specific item table (cascades); no generic bonus.
        typeStep.note = "Special → roll on " + itemTable;
        const itemStep = this.rollItemTable(itemTable, depth).root;   // shares outer budget
        return step(slot, 0, 0, label + " (Special)", null, "n/a",
          { kind: "assembly", children: [typeStep, itemStep],
            note: t1 + " Special → specific item from " + itemTable });
      }

      // Generic magic armor/weapon: type + bonus (no R3/S3 roll).
      const b = this.roller.roll(20);
      const bonusStep = this.matchOrGap(this.ds.mech[t2], b, t2, 20);
      if (slot === "weapon" && matchedType && matchedType.name !== "Sword") {
        const mb = this.ds.find(this.ds.mech[t2], b);                 // S2 Wpn Adj for non-swords
        if (mb && mb.wpn_adj) { bonusStep.label = mb.wpn_adj; bonusStep.note = "Wpn Adj (non-sword)"; }
      }
      return step(slot, 0, 0, label, null, "n/a",
        { kind: "assembly", children: [typeStep, bonusStep] });
    }

    bonusPlus(mechKey, name) {
      if (mechKey === "S2") return name.trim();
      const m = /AC Adj\s*([+-]?\d+)/.exec(name);
      if (!m) return name;
      const g = m[1];
      return (g[0] !== "+" && g[0] !== "-") ? "+" + g : g;
    }

    // ---- headline derivation --------------------------------------------
    headline(root) {
      const base = this.primaryName(root);
      if (root.table === "artifact" || root.table === "hoard") return base;
      const combines = EM.countKinds(root, new Set(["reroll", "enhancement"]));
      const ench = EM.countKinds(root, new Set(["enhancement"]));
      if (combines === 0) return base;
      let suffix = "  ·  +" + combines + " combined";
      if (ench) suffix += " (" + ench + " enchanted)";
      return base + suffix;
    }

    primaryName(root) {
      if (root.table === "armor" || root.table === "weapon") return this.assembledName(root);
      if (root.table === "master" && root.children.length) return this.primaryName(root.children[0]);
      return root.label;
    }

    assembledName(root) {
      const typeStep = root.children[0], second = root.children[1];
      if (second.table === "R2" || second.table === "S2") {   // generic: "{bonus} {type}"
        const plus = this.bonusPlus(second.table, second.label);
        return (plus + " " + typeStep.label).trim();
      }
      return this.primaryName(second);                         // Special: the R3/S3 item is the result
    }
  };
})();
